keps/sig-storage/5381-mutable-pv-affinity/README.md

```diff
@@ -0,0 +1,997 @@
+```
+The updated affinity would be:
+```yaml
+required:
+  nodeSelectorTerms:
+  - matchExpressions:
+    - key: topology.kubernetes.io/region
+      operator: In
+      values:
+      - cn-beijing
+```
+
+Or in the view of CSI accessibility requirement, from:
+```json
+[{"topology.kubernetes.io/zone": "cn-beijing-g"}]
+```
+to:
+```json
+[{"topology.kubernetes.io/region": "cn-beijing"}]
+```
+
+#### Story 2
+
+As a cluster operator, I'm conducting an upgrade to new storage category provided by our storage provider.
+However, once upgraded, the volume cannot be attached to some legacy nodes in the cluster.
+I need a way to convey this new requirement to the scheduler,
+so that my pod will not getting stuck in a `ContainerCreating` state.
+
+In this case, the old affinity would be:
+```yaml
+required:
+  nodeSelectorTerms:
+  - matchExpressions:
+    - key: provider.com/disktype.cloud_ssd
+      operator: In
+      values:
+      - available
+```
+The updated affinity would be:
+```yaml
+required:
+  nodeSelectorTerms:
+  - matchExpressions:
+    - key: provider.com/disktype.cloud_essd
+      operator: In
+      values:
+      - available
+```
+
+Or in the view of CSI accessibility requirement, from:
+```json
+[{"provider.com/disktype.cloud_ssd": "available"}]
+```
+to:
+```json
+[{"provider.com/disktype.cloud_essd": "available"}]
+```
+
+Type A node only supports cloud_ssd, while Type B node supports both cloud_ssd and cloud_essd.
+We will only allow the modification if the volume is attached to type B nodes.
+And I want to make sure the Pods using new cloud_essd volume not to be scheduled to type A nodes.
+
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+`PersistentVolume.spec.nodeAffinity.required` should works like
+`Pod.spec.affinity.nodeAffinity.requiredDuringSchedulingIgnoredDuringExecution`.
+It is never re-evaluated when the pod is already running.
+It is storage provider's responsibility to ensure that the running workload is not interrupted.
+
+**Possible race condition**
+
+There is a race condition between volume modification and pod scheduling:
+1. User updates PVC VolumeAttributeClassName.
+2. external-resizer initiate ControllerModifyVolume.
+3. User creates a new Pod and scheduler schedules it with the old affinity.
+4. external-resizer sets a new affinity to the PV.
+5. KCM/external-attacher attaches the volume to the node, and find the affinity mismatch.
+
+If this happens, the pod will be stuck in a `ContainerCreating` state.
+User will have to manually delete the pod, or using Kubernetes [descheduler](https://github.com/kubernetes-sigs/descheduler) or similar.
+
+We propose to added an `in_progress` field in `ControllerModifyVolumeResponse` to allow more timely affinity update,
```

Comment:
Before altering the topology to an unfeasible state, consider adding clearer guidance on handling scheduled pods.